File: artifacts/opl/decompose/src/opl_emu_opdata_cache.c
struct opl_emu_opdata_cache
{
  const uint16_t *waveform;
  uint32_t phase_step;
  uint32_t total_level;
  uint32_t block_freq;
  int32_t detune;
  uint32_t multiple;
  uint32_t eg_sustain;
  uint8_t eg_rate[OPL_EMU_EG_STATES];
  uint8_t eg_shift;
}
/*** DEPENDENCIES:
***/

File: artifacts/opl/translate/src/opl_emu_opdata_cache.rs
use crate::*;

#[derive(Clone)]
pub struct OplEmuOpdataCache {
    pub waveform: Option<Vec<u16>>,
    pub phase_step: u32,
    pub total_level: u32,
    pub block_freq: u32,
    pub detune: i32,
    pub multiple: u32,
    pub eg_sustain: u32,
    pub eg_rate: [u8; 4],  // Assuming OPL_EMU_EG_STATES is 4 (Attack, Decay, Sustain, Release)
    pub eg_shift: u8,
}

